File: C++/0030. 字符串重排.cpp
// we have defined the necessary header files here for this problem.
// If additional header files are needed in your program, please import here.
#include<map>
#include<vector>
int main()
{  
  // please define the C++ input here. For example: int a,b; cin>>a>>b;;  
  // please finish the function body here.  
  // please define the C++ output here. For example:cout<<____<<endl; 
    string s;
    while(cin>>s){
        map<char,int> letter;
        vector<int> fac(s.size()+1,1);
        
        for(int i=2;i<fac.size();++i){
            fac[i] = i*fac[i-1];
        }
        
        for(auto c:s){
            if(letter.count(c)){
                letter[c] ++;
            }
            else{
                letter[c] = 1;
            }
        }
        int res = fac.back();
        for(map<char,int>::iterator iter=letter.begin();iter!=letter.end();++iter){
            res /= fac[iter->second];
        }
        cout<<res<<endl;
    }
    return 0;
}

File: Python/0030. 字符串重排.py
# If you need to import additional packages or classes, please import here.
import collections
def func():
    # please define the python3 input here. 
    # For example: a,b = map(int, input().strip().split())
    # please finish the function body here.
    # please define the python3 output here. For example: print().
    while True:
        try:
            s = input().strip()
            l = len(s)
            arr = [1]*(l+1)
            for i in range(2,l+1):
                arr[i] = i*arr[i-1]
            letter = collections.defaultdict(int)
            for c in s:
                letter[c] += 1
            res = arr[l]
            for num in letter.values():
                res //= arr[num]
            print(res)

        except EOFError:
            break
    
if __name__ == "__main__":
    func()
